main.c

```diff
@@ -0,0 +1,440 @@
+    json_object *path = json_object_object_get(iso, "path");
+    json_object *size = json_object_object_get(iso, "size");
+
+    choices_t *choices = choices_create(2);
+    choices->values[0] = iso_data_create(
+            strdup("Ubuntu Server 22.10 (Kinetic Kudu)"),
+            strdup("https://releases.ubuntu.com/kinetic/ubuntu-22.10-live-server-amd64.iso"),
+            1642631168);
+    choices->values[1] = iso_data_create(
+            saprintf("Ubuntu Server %s (%s)",
+                    json_object_get_string(title),
+                    json_object_get_string(codename)),
+            saprintf("http://cdimage.ubuntu.com/%s",
+                    json_object_get_string(path)),
+            json_object_get_int(size));
+
+    json_object_put(root);
+    return choices;
+}
+
+int main(int argc, char **argv)
+{
+    args_t *args = args_create(argc, argv);
+    setlocale(LC_ALL, "C.UTF-8");
+
+    if(!initscr()) {
+        logger("initscr failure");
+        return 1;
+    }
+
+    noecho();
+
+    if(!has_colors()) {
+        logger("has_colors failure");
+        endwin();
+        return 1;
+    }
+
+    if(start_color() == ERR) {
+        logger("start_color failure");
+        endwin();
+        return 1;
+    }
+
+    keypad(stdscr, TRUE);
+
+    cbreak();
+
+    curs_set(0); /* hide */
+    refresh();
+
+    logger("start");
+
+    logger("can_change_color [%d]", can_change_color());
+    if(can_change_color()) {
+        init_color_from_bytes(ubuntu_orange, 0xE9, 0x54, 0x20);
+        init_color_from_bytes(text_white, 0xFF, 0xFF, 0xFF);
+        init_color_from_bytes(back_green, 0x0E, 0x84, 0x20);
+    } else {
+        /* These are terminal 256 color codes, see
+         * https://www.ditig.com/256-colors-cheat-sheet for an example. */
+        ubuntu_orange = 202;  /* not really but kinda close */
+        text_white = 231;
+        back_green = 28;
+    }
+
+    choices_t *iso_info = read_iso_choices(args->infile);
+
+    bool continuing = true;
+    int ch = 0;
+
+    while(continuing) {
+        orange_banner("Choose an Ubuntu version to install");
+        add_chooser(iso_info, iso_info->cur);
+        ch = getch();
+        logger("got input [%d]\n", ch);
+        switch(ch) {
+            case KEY_DOWN:
+                choice_handle_event(args, iso_info, INCREASE);
+                break;
+            case KEY_UP:
+                choice_handle_event(args, iso_info, DECREASE);
+                break;
+            case '\f':  /* ctrl-L */
+                redrawwin(stdscr);
+                break;
+            case KEY_ENTER:
+            case '\r':
+            case '\n':
+            case ' ':
+                choice_handle_event(args, iso_info, SELECT);
+                continuing = false;
+                break;
+            default:
+                break;
+        }
+    }
+
+    choices_free(iso_info);
+    args_free(args);
+
+    erase();
+    refresh();
+    endwin();
```

Comment:
I wonder if we should pass these calls to atexit to avoid exiting mid execution with a broken terminal